jni/lib/src/jni_class.dart

```diff
@@ -0,0 +1,160 @@
+import 'dart:ffi';
+
+import 'package:ffi/ffi.dart';
+
+import 'third_party/jni_bindings_generated.dart';
+import 'extensions.dart';
+import 'jvalues.dart';
+import 'jni_exceptions.dart';
+import 'jni_object.dart';
+
+part 'jni_class_methods_generated.dart';
+
+final ctorLookupChars = "<init>".toNativeChars();
+
+/// Convenience wrapper around a JNI local class reference.
+///
+/// Reference lifetime semantics are same as [JniObject].
+class JniClass {
+  final JClass _cls;
+  final Pointer<JniEnv> _env;
+  bool _deleted = false;
+  JniClass.of(this._env, this._cls);
+
+  JniClass.fromJClass(Pointer<JniEnv> env, JClass cls)
+      : _env = env,
+        _cls = cls;
+
+  JniClass.fromGlobalRef(Pointer<JniEnv> env, JniGlobalClassRef r)
+      : _env = env,
+        _cls = env.NewLocalRef(r._cls) {
```

Comment:
This is actually what Daco & I were discussing. There are quite a few edge cases like this, and it makes sense to use Global references entirely, and also provide a thread-safe indirection over JniEnv if possible.

> My understanding is that local refs only last as long as the current native scope. Maybe this doesn't matter too much since the outermost scope will be Dart. But I think it might matter in cases where the user passes a Dart callback to Java, and Java calls back into Dart?

I don't know how to verify the following
>> One unusual case deserves separate mention. If you attach a native thread with AttachCurrentThread, the code you are running will never automatically free local references until the thread detaches. Any local references you create will have to be deleted manually. In general, any native code that creates local references in a loop probably needs to do some manual deletion.

https://developer.android.com/training/articles/perf-jni

But the concern is valid. On Android, thread is attached manually so far in my experimentation, but there may be cases it isn't.